python_modules/libraries/dagster-dg-cli/dagster_dg_cli/cli/api/API_CONVENTIONS.md

```diff
@@ -0,0 +1,302 @@
+# Dagster Plus API CLI Conventions
+
+This document outlines the conventions for implementing CLI commands in the `dg api` namespace, following GitHub CLI best practices.
+
+## Command Structure
+
+The API commands follow the pattern: `dg api <noun> <verb>`
+
+### Examples:
+
+```bash
+dg api deployment list
+dg api deployment view <id>
+dg api secret list
+dg api secret create
+dg api agent list
+```
+
+This mirrors GitHub CLI patterns like:
+
+- `gh repo list`
+- `gh issue create`
+- `gh pr view`
+
+## File Organization
+
+### Directory Structure:
+
+```
+dagster_dg_cli/cli/plus/api/
+├── __init__.py                 # Main API group
+├── deployment.py               # Deployment noun with all verbs
+├── secret.py                   # Secret noun with all verbs
+├── agent.py                    # Agent noun with all verbs
+├── check.py                    # Asset check noun with all verbs
+├── code_location.py            # Code location noun with all verbs
+├── env_var.py                  # Environment variable noun with all verbs
+└── shared.py                   # Shared utilities
+```
+
+### Implementation Pattern:
+
+Each noun file (e.g., `deployment.py`) contains:
+
+1. **Noun group** - Click group for the resource
+2. **Verb commands** - Individual commands for operations
+3. **Helper functions** - Shared utilities for that noun
+
+## Standard Flags
+
+### Required for ALL API Commands:
+
+- `--json` flag for machine-readable JSON output
+- Human-readable table format as default
+
+### Common Optional Flags:
+
+- `--limit <n>` for pagination
+- `--filter <query>` for filtering results
+- `--format <table|json>` (alternative to --json flag)
+
+## Output Formatting
+
+### Human-Readable (Default):
+
+```
+Name: my-deployment
+ID: abc-123
+Type: SERVERLESS
+
+Name: other-deployment
+ID: def-456
+Type: HYBRID
+```
+
+### JSON Format (--json flag):
+
+```json
+[
+  {
+    "name": "my-deployment",
+    "id": "abc-123",
+    "type": "SERVERLESS"
+  },
+  {
+    "name": "other-deployment",
+    "id": "def-456",
+    "type": "HYBRID"
+  }
+]
+```
+
+## Error Handling
+
+### Human-Readable Errors:
+
+```
+Error querying Dagster Plus API: Unauthorized access
+```
+
+### JSON Errors (when --json used):
+
+```json
+{
+  "error": "Unauthorized access"
```

Comment:
do we want error codes? I could imagine clients wanting to handle auth errors in a generic way without parsing human-readable strings. We could just use http status codes for that too.

Reply:
This is a good idea but a big enough one to warrant a standalone followup